File: frontend/src/engine/cupMesh.ts
import { catmullRomSpline, scaleControlPoints, type Point2D } from './cupCurve'

export interface CupMeshParams {
  controlPoints: Point2D[]
  centerX: number
  wallThickness: number
  cupHeight: number
  baseHeight?: number
  thetaSegments?: number
  profileSamples?: number
}

export interface CupMeshData {
  positions: Float32Array
  indices: Uint32Array
  uvs: Float32Array
  /** 外壁顶点的真实曲面法线（与 positions 前 outerVertexCount 个顶点一一对应） */
  outerNormals: Float32Array
  yMin: number
  yMax: number
  outerVertexCount: number
  /** 外壁三角面在 indices 中的数量（用于多材质分组） */
  outerFaceCount: number
  profileRowCount: number
  /** 环向分段数（顶点列数 = ringSegments + 1，含闭合接缝） */
  ringSegments: number
}

export function buildCupMesh(params: CupMeshParams): CupMeshData {
  const {
    controlPoints,
    centerX,
    wallThickness,
    cupHeight,
    baseHeight = 350,
    thetaSegments = 120,
    profileSamples = 180,
  } = params

  const scaled = scaleControlPoints(controlPoints, cupHeight, baseHeight)
  const curve = catmullRomSpline(scaled, profileSamples)

  const outerR = curve.map((p) => Math.abs(p.x - centerX))
  const innerR = outerR.map((r) => Math.max(2, r - wallThickness))

  // Three.js 使用 Y 轴朝上：杯底 y=0，杯口朝上
  const profileBottomY = Math.max(...curve.map((p) => p.y))
  const heights = curve.map((p) => profileBottomY - p.y)
  const innerFloorY = heights[heights.length - 1]
  const outerBottomY = innerFloorY - wallThickness
  const yMin = outerBottomY
  const yMax = Math.max(...heights)
  const yRange = Math.max(1, yMax - yMin)

  // 子午线（半径, 高度）平面内的外壁法线，逐行计算后旋转得到 3D 曲面法线
  const meridianN: { nr: number; nh: number }[] = []
  for (let i = 0; i < heights.length; i++) {
    const ip = Math.max(0, i - 1)
    const iq = Math.min(heights.length - 1, i + 1)
    const dR = outerR[iq] - outerR[ip]
    const dH = heights[iq] - heights[ip]
    // 切线 (dR, dH) 的法线，取径向分量为正（指向杯外）
    let nr = dH
    let nh = -dR
    if (nr < 0) {
      nr = -nr
      nh = -nh
    }
    const len = Math.hypot(nr, nh) || 1
    meridianN.push({ nr: nr / len, nh: nh / len })
  }

  const verts: number[] = []
  const uvs: number[] = []
  const outerNorm: number[] = []
  const faces: number[] = []
  const outerIdx: number[][] = []
  const innerIdx: number[][] = []

  const ringSegments = thetaSegments
  const cols = ringSegments + 1

  for (let i = 0; i < heights.length; i++) {
    const row: number[] = []
    const { nr, nh } = meridianN[i]
    const yOuter = i === heights.length - 1 ? outerBottomY : heights[i]
    for (let j = 0; j <= ringSegments; j++) {
      const theta = (j / ringSegments) * Math.PI * 2
      const cos = Math.cos(theta)
      const sin = Math.sin(theta)
      const vx = outerR[i] * cos
      const vz = outerR[i] * sin
      verts.push(vx, yOuter, vz)
      uvs.push(j / ringSegments, (yOuter - yMin) / yRange)
      outerNorm.push(nr * cos, nh, nr * sin)
      row.push((verts.length / 3) - 1)
    }
    outerIdx.push(row)
  }

  for (let i = 0; i < heights.length; i++) {
    const row: number[] = []
    const yInner = i === heights.length - 1 ? innerFloorY : heights[i]
    for (let j = 0; j <= ringSegments; j++) {
      const theta = (j / ringSegments) * Math.PI * 2
      const vx = innerR[i] * Math.cos(theta)
      const vz = innerR[i] * Math.sin(theta)
      verts.push(vx, yInner, vz)
      uvs.push(j / ringSegments, (yInner - yMin) / yRange)
      row.push((verts.length / 3) - 1)
    }
    innerIdx.push(row)
  }

  // 环向闭合：最后一列与第一列同位置（θ=2π ≡ 0）
  for (let i = 0; i < heights.length - 1; i++) {
    for (let j = 0; j < ringSegments; j++) {
      const a = outerIdx[i][j]
      const b = outerIdx[i][j + 1]
      const c = outerIdx[i + 1][j]
      const d = outerIdx[i + 1][j + 1]
      faces.push(a, b, c, b, d, c)
    }
  }

  const outerFaceCount = faces.length

  for (let i = 0; i < heights.length - 1; i++) {
    for (let j = 0; j < ringSegments; j++) {
      const a = innerIdx[i][j]
      const b = innerIdx[i + 1][j]
      const c = innerIdx[i][j + 1]
      const d = innerIdx[i + 1][j + 1]
      faces.push(a, b, c, c, b, d)
    }
  }

  // 杯口壁厚环（法线朝上，连接内外壁，像真实杯沿）
  const topOuter = outerIdx[0]
  const topInner = innerIdx[0]
  for (let j = 0; j < ringSegments; j++) {
    const a = topOuter[j]
    const b = topInner[j]
    const c = topOuter[j + 1]
    const d = topInner[j + 1]
    faces.push(a, b, c, b, d, c)
  }

  const bottomOuter = outerIdx[heights.length - 1]
  const bottomInner = innerIdx[heights.length - 1]

  // 杯底壁厚环：内底缘（innerFloorY）→ 外底缘（outerBottomY），竖直厚度 = wallThickness
  for (let j = 0; j < ringSegments; j++) {
    const a = bottomInner[j]
    const b = bottomInner[j + 1]
    const c = bottomOuter[j + 1]
    const d = bottomOuter[j]
    faces.push(a, b, c, a, c, d)
  }

  const innerCenterIndex = verts.length / 3
  verts.push(0, innerFloorY, 0)
  uvs.push(0.5, 1)

  const outerCenterIndex = verts.length / 3
  verts.push(0, outerBottomY, 0)
  uvs.push(0.5, 0)

  // 内底面（法线朝上，从杯内可见）
  for (let j = 0; j < ringSegments; j++) {
    faces.push(innerCenterIndex, bottomInner[j + 1], bottomInner[j])
  }

  // 外底面（法线朝下，从杯外/桌面方向可见）
  for (let j = 0; j < ringSegments; j++) {
    faces.push(outerCenterIndex, bottomOuter[j], bottomOuter[j + 1])
  }

  return {
    positions: new Float32Array(verts),
    indices: new Uint32Array(faces),
    uvs: new Float32Array(uvs),
    outerNormals: new Float32Array(outerNorm),
    yMin,
    yMax,
    outerVertexCount: heights.length * cols,
    outerFaceCount,
    profileRowCount: heights.length,
    ringSegments,
  }
}

File: backend/pipeline/cup_mesh.py
"""
杯型回转体网格生成 — 由 thr cup3.py 提炼，用于 Web API 导出。
"""

from __future__ import annotations

import io

import numpy as np
import trimesh
from scipy.interpolate import splprep, splev

import config


def scale_control_points(
    points: list[tuple[float, float]],
    cup_height: float,
    base_height: float = config.CUP_BASE_HEIGHT,
    top_y: float = 80.0,
) -> list[tuple[float, float]]:
    scale = cup_height / base_height
    return [(x, top_y + (y - top_y) * scale) for x, y in points]


def profile_curve(
    control_points: list[tuple[float, float]],
    center_x: float = config.CUP_CENTER_X,
    samples: int = config.CUP_PROFILE_SAMPLES,
) -> tuple[np.ndarray, np.ndarray]:
    pts = np.array(control_points, dtype=np.float64)
    x = pts[:, 0]
    y = pts[:, 1]
    tck, _ = splprep([x, y], s=0)
    unew = np.linspace(0, 1, samples)
    out = splev(unew, tck)
    return np.array(out[0]), np.array(out[1])


def build_cup_mesh(
    control_points: list[tuple[float, float]],
    *,
    center_x: float = config.CUP_CENTER_X,
    wall_thickness: float = config.CUP_WALL_THICKNESS_DEFAULT,
    cup_height: float = config.CUP_HEIGHT_DEFAULT,
    theta_segments: int = config.CUP_REVOLVE_SEGMENTS,
) -> trimesh.Trimesh:
    scaled = scale_control_points(control_points, cup_height)
    x, y = profile_curve(scaled, center_x=center_x)

    outer_r = np.abs(x - center_x)
    inner_r = np.clip(outer_r - wall_thickness, 2, None)
    z = 700 - y

    inner_floor_z = float(z[-1])
    outer_bottom_z = inner_floor_z - wall_thickness

    # endpoint=False：首尾角度不重复，再用取模闭合侧面
    theta = np.linspace(0, 2 * np.pi, theta_segments, endpoint=False)

    verts: list[list[float]] = []
    faces: list[list[int]] = []
    outer_idx: list[list[int]] = []
    inner_idx: list[list[int]] = []

    last = len(z) - 1
    for i in range(len(z)):
        row: list[int] = []
        z_outer = outer_bottom_z if i == last else float(z[i])
        for t in theta:
            verts.append([
                float(outer_r[i] * np.cos(t)),
                float(outer_r[i] * np.sin(t)),
                z_outer,
            ])
            row.append(len(verts) - 1)
        outer_idx.append(row)

    for i in range(len(z)):
        row = []
        z_inner = inner_floor_z if i == last else float(z[i])
        for t in theta:
            verts.append([
                float(inner_r[i] * np.cos(t)),
                float(inner_r[i] * np.sin(t)),
                z_inner,
            ])
            row.append(len(verts) - 1)
        inner_idx.append(row)

    cols = len(theta)

    for i in range(len(z) - 1):
        for j in range(cols):
            j1 = (j + 1) % cols
            a, b, c, d = outer_idx[i][j], outer_idx[i][j1], outer_idx[i + 1][j], outer_idx[i + 1][j1]
            faces.append([a, b, c])
            faces.append([b, d, c])

    for i in range(len(z) - 1):
        for j in range(cols):
            j1 = (j + 1) % cols
            a, b, c, d = inner_idx[i][j], inner_idx[i + 1][j], inner_idx[i][j1], inner_idx[i + 1][j1]
            faces.append([a, b, c])
            faces.append([c, b, d])

    top_outer = outer_idx[0]
    top_inner = inner_idx[0]
    for j in range(cols):
        j1 = (j + 1) % cols
        a, b, c, d = top_outer[j], top_inner[j], top_outer[j1], top_inner[j1]
        faces.append([a, c, b])
        faces.append([c, d, b])

    bottom_outer = outer_idx[-1]
    bottom_inner = inner_idx[-1]
    for j in range(cols):
        j1 = (j + 1) % cols
        a, b = bottom_inner[j], bottom_inner[j1]
        c, d = bottom_outer[j1], bottom_outer[j]
        faces.append([a, b, c])
        faces.append([a, c, d])

    inner_center = len(verts)
    verts.append([0.0, 0.0, inner_floor_z])
    outer_center = len(verts)
    verts.append([0.0, 0.0, outer_bottom_z])

    # 内底面（法线朝 +Z，杯内可见）
    for j in range(cols):
        j1 = (j + 1) % cols
        a, b = bottom_inner[j], bottom_inner[j1]
        faces.append([inner_center, a, b])

    # 外底面（法线朝 -Z，杯外/桌面方向可见）
    for j in range(cols):
        j1 = (j + 1) % cols
        a, b = bottom_outer[j], bottom_outer[j1]
        faces.append([outer_center, b, a])

    mesh = trimesh.Trimesh(
        vertices=np.array(verts, dtype=np.float64),
        faces=np.array(faces, dtype=np.int64),
        process=True,
    )
    return mesh


def mesh_to_obj_bytes(mesh: trimesh.Trimesh) -> bytes:
    buf = io.BytesIO()
    mesh.export(buf, file_type="obj")
    return buf.getvalue()


def mesh_to_stl_bytes(mesh: trimesh.Trimesh) -> bytes:
    buf = io.BytesIO()
    mesh.export(buf, file_type="stl")
    return buf.getvalue()
